### PR TITLE
 (proof-new) Witness axiom reconstruction for purification skolems

### PR DESCRIPTION
**Pull request by ajreynol**
This formalizes the proofs of existentials that justify purification variables, e.g. those with witness form `witness x. x = t` for the term `t` they purify. 

This PR generalizes EXISTS_INTRO to do this, and makes some simplifications to SkolemManager.

### CODE REVIEW

File: src/smt/witness_form.cpp [hunk 1]
```diff
@@ -92,6 +93,17 @@ Node WitnessFormGenerator::convertToWitnessForm(Node t)
...
           ProofGenerator* pg = skm->getProofGenerator(exists);
+          if (pg == nullptr)
+          {
+            // it may be a purification skolem
+            pg = convertExistsInternal(exists);
+            if (pg == nullptr)
+            {
+              Trace("witness-form")
+                  << "WitnessFormGenerator: No proof generator for " << exists
+                  << std::endl;
```
Comment (HanielB) [inline src/smt/witness_form.cpp:104]: Don't we want to use some kind of default proof here?

File: src/theory/quantifiers/proof_checker.cpp [hunk 2]
```diff
@@ -43,8 +44,25 @@ Node QuantifiersProofRuleChecker::checkInternal(
...
+      return Node::null();
+    }
+    // substitution must contain only the variable of the existential
+    for (const std::pair<const Node, Node>& s : subs)
+    {
+      if (s.first != exists[0][0])
+      {
+        return Node::null();
+      }
+    }
```
Comment (HanielB) [inline src/theory/quantifiers/proof_checker.cpp:64]: maybe check this first? The match should be the most expensive check.

Comment (ajreynol) [reply src/theory/quantifiers/proof_checker.cpp:64]: `subs` is computed by the match :)

File: src/smt/witness_form.cpp (hunk 1 above)
Comment (ajreynol) [reply src/smt/witness_form.cpp:104]: Yes, this is just a warning, we fill in via WITNESS_AXIOM below when none is provided.

File: src/theory/quantifiers/proof_checker.cpp (hunk 2 above)
Comment (HanielB) [reply src/theory/quantifiers/proof_checker.cpp:64]: Indeed that makes sense! :)

File: src/smt/witness_form.cpp (hunk 1 above)
Comment (HanielB) [reply src/smt/witness_form.cpp:104]: I mean what is gonna be the justification of the existential? Do I understand correctly that it'll be an open leaf?

Comment (ajreynol) [reply src/smt/witness_form.cpp:104]: It's justified by a trusted rule "WITNESS_AXIOM".

Comment (HanielB) [reply src/smt/witness_form.cpp:104]: Ah, I see. I forgot about this feature of `addLazyStep`. Maybe add a comment to clarify this?

**Review by HanielB — APPROVED** [review]

**Pull request merged**